cli/ml_training.go

```diff
@@ -597,3 +613,283 @@ func convertVisibilityToProto(visibility string) (*v1.Visibility, error) {
+	}
+
+	return nil
+}
+
+// createTrainingScript creates the shell script content to run inside the container.
+func createTrainingScript(customArgs []string) (string, error) {
+	// Validate custom arguments format (key=value) before building script
+	for _, arg := range customArgs {
+		if !strings.Contains(arg, "=") {
+			return "", errors.Errorf("invalid custom argument format: %s (expected key=value)", arg)
+		}
+
+		// Validate that the key portion only contains safe characters
+		parts := strings.SplitN(arg, "=", 2)
+		key := parts[0]
+		if !isValidArgumentKey(key) {
+			return "", errors.Errorf("invalid argument key: %s (only alphanumeric characters, underscores, and hyphens are allowed)", key)
+		}
+	}
+
+	var script strings.Builder
+
+	// Script header and setup
+	script.WriteString("#!/bin/bash\n")
+	script.WriteString("set -e\n\n")
+	script.WriteString("echo \"Installing training script package...\"\n")
+	script.WriteString("pip3 install --no-cache-dir .\n\n")
+	script.WriteString("echo \"Running training...\"\n")
+
+	// Build Python training command
+	script.WriteString("python3 -m model.training")
+	script.WriteString(" --dataset_file=/dataset.jsonl")
+	script.WriteString(" --model_output_directory=/model_output")
+
+	// Add custom arguments
+	for _, arg := range customArgs {
+		parts := strings.SplitN(arg, "=", 2)
+		key := parts[0]
+		value := parts[1]
+
+		script.WriteString(" --")
+		script.WriteString(key)
+		script.WriteString("=")
+		// Use strconv.Quote for proper shell quoting - it returns a double-quoted string
+		// with all special characters properly escaped
+		script.WriteString(strconv.Quote(value))
+	}
+	script.WriteString("\n\n")
+
+	// Script footer
+	script.WriteString("echo \"Training completed successfully!\"\n")
+
+	return script.String(), nil
+}
+
+// isValidArgumentKey validates that an argument key only contains safe characters.
+// Allowed characters: letters (a-z, A-Z), digits (0-9), underscores (_), and hyphens (-).
+func isValidArgumentKey(key string) bool {
+	if key == "" {
+		return false
+	}
+
+	for _, char := range key {
+		if !((char >= 'a' && char <= 'z') ||
+			(char >= 'A' && char <= 'Z') ||
+			(char >= '0' && char <= '9') ||
+			char == '_' ||
+			char == '-') {
+			return false
+		}
+	}
+
+	return true
+}
+
+// getContainerImageURI returns the full container image URI based on the version.
+func getContainerImageURI(c *viamClient, version string) (string, error) {
+	res, err := c.mlTrainingClient.ListSupportedContainers(context.Background(), &mltrainingpb.ListSupportedContainersRequest{})
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to list supported containers")
+	}
+
+	containerKeyList := []string{}
+	for key := range res.ContainerMap {
+		containerKeyList = append(containerKeyList, key)
+	}
+	slices.Sort(containerKeyList)
+
+	container, ok := res.ContainerMap[version]
+	if !ok {
+		if dockerVertexImageRegex.MatchString(version) {
```

Comment:
There's no feature flags to use here, so I threw this back in, but lmk if I should remove.

Reply:
Maybe we expose an endpoint for `GetContainer()` that returns a containerURI given a key (and would just return the URI if they're a power user? @tahiyasalam what do you think?

Reply:
added the warning, thanks!